JavaScript/Angular.md

```diff
@@ -0,0 +1,1551 @@
+
+<!-- bad -->
+<my-app-component (onSomethingHappened)="onSomethingHappened($event)">
+
+<!-- good -->
+<my-app-component (somethingHappened)="onSomethingHappened($event)">
+```
+
+----
+
+#### Name click handlers in a descriptive manner:
+
+```html
+<!-- bad -->
+<button (click)="onClick()">Log in</button>
+
+<!-- bad - use infinitive instead of past form -->
+<button (click)="onLogInClicked()">Log in</button>
+
+<!-- bad - too much -->
+<button (click)="onLogInButtonClick()">Log in</button>
+
+<!-- good -->
+<button (click)="onLogInClick()">Log in</button>
+```
+
+----
+
+#### Use safe navigation operator:
+
+```html
+<!-- bad -->
+<div *ngIf="user && user.loggedIn"></div>
+
+<!-- good -->
+<div *ngIf="user?.loggedIn"></div>
+```
+
+----
+
+#### Add space around interpolation curly brackets:
+
+```html
+<!-- bad -->
+<div>{{userName}}</div>
+
+<!-- good -->
+<div>{{ userName }}</div>
+```
+
+----
+
+#### Use `ng-container` when possible to reduce the amount of generated DOM elements:
+
+```html
+<!-- bad -->
+<div *ngIf="something"></div>
+
+<!-- good -->
+<ng-container *ngIf="something"></ng-container>
+
+<!-- good -->
+<div class="i-need-this-class" *ngIf="something"></div>
+```
+
+----
+
+#### Use `*ngIf; else`:
+
+```html
+<!-- bad -->
+<ng-container *ngIf="something"></ng-container>
+<ng-container *ngIf="!something"></ng-container>
+
+<!-- good -->
+<ng-container *ngIf="something; else anotherThing"></ng-container>
+<ng-template #anotherThing></ng-template>
+```
+
+----
+
+#### Subscribe to observables with `async` pipe if possible:
+
+Component:
+
+```typescript
+@Component(...)
+export class ArticlesList {
+  public articles$: Observable<Array<Article>> = this.articleService.fetchArticles();
+  public menu$: Observable<Menu> = this.articleService.fetchMenu();
+  ...
+}
+```
+
+Template:
+
+```html
+<my-app-menu
+  *ngIf="menu$ | async as articlesMenu"
+  [menu]="articlesMenu"
+></my-app-menu>
+
+<my-app-article-details
+  *ngFor="let article of articles$ | async"
+  [article]="article"
+></my-app-article-details>
+```
+
+`async` pipe is especially useful if `changeDetection` is `OnPush` since it subscribes, calls change detection on changes and unsubscribes when component is destroyed.
+
+----
+
+#### Use attributes for transclusion selectors:
+
+`my-app-wrapper` component template:
+
+```html
+<div class="wrapper">
+  <p>Wrapper content:</p>
+  <ng-content select="[some-stuff]"></ng-content>
+</div>
+```
+
+Some other component's template:
+
+```html
+<my-app-wrapper>
+  <div class="some-item" some-stuff>Some stuff</div>
+</my-app-wrapper>
+```
+
+----
+
+#### Order attributes and bindings:
+
+1. Structural directives (`*ngFor`, `*ngIf`, etc.)
+2. Element reference (`#myComponent`)
+3. Ordinary attributes (`class`, etc.)
+4. Non-interpolated string inputs (`foo="bar"`)
+5. Interpolated inputs (`[foo]="theBar"`)
+6. Two-way bindings (`[(ngModel)]="email"`)
+7. Outputs (`(click)="onClick()"`)
```

Comment:
We can check what is possible with https://github.com/mgechev/codelyzer

Reply:
We're missing the animation trigger. I'm guessing they can come last?

Reply:
Added them after structural directives since they are usually relevant when components show/hide. But we do not use them that much anyway and honestly I do not care that much where they are placed :D